### PR TITLE
K8SPG-80 Wait for pod termination before restore

### PR DESCRIPTION
**Pull request by nonemax**

### CODE REVIEW

**Comment by JNKPercona** [comment]
| Test name  | Status |
| ------------- | ------------- |
| init-deploy | passed |
| demand-backup | failed |
| data-migration-gcs | passed |
| scaling | passed |
| recreate | passed |
| affinity | passed |
| monitoring | passed |

commit: https://github.com/percona/percona-postgresql-operator/pull/82/commits/f785fe5312158783c0c41a6853f56a4227bd7ca3
image: `perconalab/percona-postgresql-operator:PR-82-f785fe53-pgo-apiserver`

image: `perconalab/percona-postgresql-operator:PR-82-f785fe53-pgo-event`

image: `perconalab/percona-postgresql-operator:PR-82-f785fe53-pgo-rmdata`

image: `perconalab/percona-postgresql-operator:PR-82-f785fe53-pgo-scheduler`

image: `perconalab/percona-postgresql-operator:PR-82-f785fe53-postgres-operator`

image: `perconalab/percona-postgresql-operator:PR-82-f785fe53-pgo-deployer`

**Comment by JNKPercona** [comment]
| Test name  | Status |
| ------------- | ------------- |
| init-deploy | passed |
| demand-backup | failed |
| data-migration-gcs | passed |
| scaling | passed |
| recreate | passed |
| affinity | passed |
| monitoring | passed |

commit: https://github.com/percona/percona-postgresql-operator/pull/82/commits/14994866a5f410a538402445a97cbfc65030bf68
image: `perconalab/percona-postgresql-operator:PR-82-14994866-pgo-apiserver`

image: `perconalab/percona-postgresql-operator:PR-82-14994866-pgo-event`

image: `perconalab/percona-postgresql-operator:PR-82-14994866-pgo-rmdata`

image: `perconalab/percona-postgresql-operator:PR-82-14994866-pgo-scheduler`

image: `perconalab/percona-postgresql-operator:PR-82-14994866-postgres-operator`

image: `perconalab/percona-postgresql-operator:PR-82-14994866-pgo-deployer`

**Comment by JNKPercona** [comment]
| Test name  | Status |
| ------------- | ------------- |
| init-deploy | passed |
| demand-backup | passed |
| data-migration-gcs | passed |
| scaling | passed |
| recreate | passed |
| affinity | passed |
| monitoring | passed |

commit: https://github.com/percona/percona-postgresql-operator/pull/82/commits/a90d81267862a5d982a113362e0002aad5c76a7a
image: `perconalab/percona-postgresql-operator:PR-82-a90d8126-pgo-apiserver`

image: `perconalab/percona-postgresql-operator:PR-82-a90d8126-pgo-event`

image: `perconalab/percona-postgresql-operator:PR-82-a90d8126-pgo-rmdata`

image: `perconalab/percona-postgresql-operator:PR-82-a90d8126-pgo-scheduler`

image: `perconalab/percona-postgresql-operator:PR-82-a90d8126-postgres-operator`

image: `perconalab/percona-postgresql-operator:PR-82-a90d8126-pgo-deployer`

File: e2e-tests/demand-backup/run [hunk 1]
```diff
@@ -90,24 +91,23 @@ function main {
...
+
 
 	desc 'Testing backup/restore from local'
 	apply_configs ${cluster}
 	spinup_pgcluster "${cluster}" "${src_dir}/deploy/cr.yaml" "local" "${cluster}-custom-config"
 	prepare_n_restore ${cluster} "posix" "full"
 
 	kubectl_bin delete perconapgcluster/${cluster}
-	sleep 30
+	sleep 60
```
Comment (cap1984) [inline e2e-tests/demand-backup/run:102]: I believe we need to do something with 1-2 minute freeze inside operator

Comment (nonemax) [reply e2e-tests/demand-backup/run:102]: As I understand operator didn't freeze but pod itself

**Comment by JNKPercona** [comment]
| Test name  | Status |
| ------------- | ------------- |
| init-deploy | passed |
| demand-backup | passed |
| data-migration-gcs | passed |
| scaling | passed |
| recreate | passed |
| affinity | passed |
| monitoring | passed |

commit: https://github.com/percona/percona-postgresql-operator/pull/82/commits/033c87c3b1e8aa8ec63695aa8fea7c23ce2c2187
image: `perconalab/percona-postgresql-operator:PR-82-033c87c3-pgo-apiserver`

image: `perconalab/percona-postgresql-operator:PR-82-033c87c3-pgo-event`

image: `perconalab/percona-postgresql-operator:PR-82-033c87c3-pgo-rmdata`

image: `perconalab/percona-postgresql-operator:PR-82-033c87c3-pgo-scheduler`

image: `perconalab/percona-postgresql-operator:PR-82-033c87c3-postgres-operator`

image: `perconalab/percona-postgresql-operator:PR-82-033c87c3-pgo-deployer`

**Comment by JNKPercona** [comment]
| Test name  | Status |
| ------------- | ------------- |
| init-deploy | passed |
| demand-backup | passed |
| data-migration-gcs | passed |
| scaling | passed |
| recreate | passed |
| affinity | passed |
| monitoring | passed |

commit: https://github.com/percona/percona-postgresql-operator/pull/82/commits/693045c74f34f5c72eea42cba777309c9eac04d8
image: `perconalab/percona-postgresql-operator:PR-82-693045c7-pgo-apiserver`

image: `perconalab/percona-postgresql-operator:PR-82-693045c7-pgo-event`

image: `perconalab/percona-postgresql-operator:PR-82-693045c7-pgo-rmdata`

image: `perconalab/percona-postgresql-operator:PR-82-693045c7-pgo-scheduler`

image: `perconalab/percona-postgresql-operator:PR-82-693045c7-postgres-operator`

image: `perconalab/percona-postgresql-operator:PR-82-693045c7-pgo-deployer`

File: percona/controllers/pgc/pgc.go [hunk 2]
```diff
@@ -285,3 +291,52 @@ func (c *Controller) AddPerconaPGClusterEventHandler() {
 func (c *Controller) WorkerCount() int {
 	return c.PerconaPGClusterWorkerCount
 }
+
+func PrepareForRestore(clientset *kubeapi.Client, clusterName, namespace string) error {
+	ctx := context.TODO()
+	err := deleteDatabasePods(clientset, clusterName, namespace)
+	if err != nil {
+		log.Error("get pods: %s", err.Error())
```
Comment (qJkee) [inline percona/controllers/pgc/pgc.go:299]: Errorf

File: percona/controllers/pgc/pgc.go [hunk 3]
```diff
@@ -285,3 +291,52 @@ func (c *Controller) AddPerconaPGClusterEventHandler() {
...
+	return nil
+}
+
+func deleteDatabasePods(clientset *kubeapi.Client, clusterName, namespace string) error {
+	ctx := context.TODO()
+	pgInstances, err := clientset.CoreV1().Pods(namespace).List(ctx, metav1.ListOptions{
+		LabelSelector: fmt.Sprintf("%s=%s,%s", config.LABEL_PG_CLUSTER, clusterName,
+			config.LABEL_PG_DATABASE),
+	})
+	if err != nil {
```
Comment (qJkee) [inline percona/controllers/pgc/pgc.go:321]: we need to check if err there is not `NotFound` error.
Because in case of no pods, we will return error

**Review by qJkee — APPROVED** [review]

**Review by qJkee — APPROVED** [review]

**Comment by JNKPercona** [comment]
| Test name  | Status |
| ------------- | ------------- |
| demand-backup | passed |
| init-deploy | passed |
| data-migration-gcs | passed |
| scaling | passed |
| recreate | passed |
| affinity | passed |
| monitoring | passed |

commit: https://github.com/percona/percona-postgresql-operator/pull/82/commits/4108f907b3949a406cea3671fe578d09da71dffc
image: `perconalab/percona-postgresql-operator:PR-82-4108f907-pgo-apiserver`

image: `perconalab/percona-postgresql-operator:PR-82-4108f907-pgo-event`

image: `perconalab/percona-postgresql-operator:PR-82-4108f907-pgo-rmdata`

image: `perconalab/percona-postgresql-operator:PR-82-4108f907-pgo-scheduler`

image: `perconalab/percona-postgresql-operator:PR-82-4108f907-postgres-operator`

image: `perconalab/percona-postgresql-operator:PR-82-4108f907-pgo-deployer`

**Review by cap1984 — APPROVED** [review]

**Review by hors — APPROVED** [review]

**Pull request merged**